Docs: Improve format documentation (#963)
json-schema-org/website: pages/understanding-json-schema/reference/type.md

For example, because JSON doesn't have a "DateTime" type, dates need to be encoded as strings. `format` allows the schema author to indicate that the string value should be interpreted as a date. By default, `format` is just an annotation and does not affect validation.

Optionally, validator [implementations](../../learn/glossary#implementation) can provide a configuration option to enable `format` to function as an assertion rather than just an annotation. That means that validation fails when, for example, a value with a `date` format isn't in a form that can be parsed as a date. This allows values to be constrained beyond what other tools in JSON Schema, including [Regular Expressions](../../understanding-json-schema/reference/regular_expressions), can do.

> Implementations may provide validation for only a subset of the built-in formats or do partial validation for a given format. For example, some implementations may consider a string an email if it contains an `@`, while others might perform additional checks for other aspects of a well-formed email address.

> The JSON Schema specification has a bias toward networking-related formats, likely due to its roots in web technologies. However, custom formats may also be used if the parties exchanging the JSON documents share information about the custom format types. A JSON Schema validator will ignore any format type it does not understand.

Reviewer: > The JSON Schema specification has a bias toward networking-related formats, likely due to its roots in web technologies.

I don't know how important this really is.  I think it's sufficient to say that the spec defines some but you can define your own if all consumers agree what that custom format means (and they all configure their validators to behave the same).

Author: Hi @gregsdennis 

**Original:**

The JSON Schema specification has a bias toward networking-related formats, likely due to its roots in web technologies. However, custom formats may also be used if the parties exchanging the JSON documents share information about the custom format types. A JSON Schema validator will ignore any format type it does not understand.

**Revised:**

The JSON Schema specification defines some standard formats, but you can also define custom formats as long as all consumers agree on their meaning and configure their validators to handle them consistently. A JSON Schema validator will ignore any format type it does not understand.

Do you agree with this?

### Built-in Formats

> It should be noted that `format` is not limited because it only defines a specific set of valid values for formats. Users may define their own custom keywords to work with any specific data type, such as `integer`, `double`, `float`, etc. Below, we cover the commonly used `string` formats specified in the JSON Schema specification.

Reviewer 2: Formats can only apply to JSON data types which doesn't include `integer`, `double`, and `float`. Although any type can be used, the only types that would make any sense to use with `format` are `string` and `number`. I think the important thing this paragraph should be communicating is that although the JSON Schema specification only defines `string` formats, it's allowed to define `number` formats as well.

Here's my suggestion, but it could probably use a little more work.

```suggestion
It should be noted that `format` is not limited to a specific set of valid values or types. Users may define their own custom keywords including ones that work with JSON data types other than `string`, such as `number`. Below, we cover the formats specified in the JSON Schema specification.
```

#### Dates and Times

> Dates and times are represented in [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6). This is a subset of the date format also commonly known as [ISO8601 format](https://www.iso.org/iso-8601-date-and-time-format.html).

Reviewer: > This is a subset of the date format also commonly known as ISO8601 format.

This is not true.  Here's a really good comparison: https://ijmacd.github.io/rfc3339-iso8601/

Author: Hi @gregsdennis ,

I have two different suggestions for this:

**1:**
Dates and Times
Dates and times are represented in [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6). While RFC 3339 is often considered a subset of the [ISO 8601 format](https://www.iso.org/iso-8601-date-and-time-format.html), there are important differences between the two. For a detailed comparison, refer to this resource: [RFC 3339 vs. ISO 8601](https://ijmacd.github.io/rfc3339-iso8601/).

**2:**
Dates and times are represented in [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6). While RFC 3339 is often seen as a simpler version of the [ISO 8601 format](https://www.iso.org/iso-8601-date-and-time-format.html), there are some differences. RFC 3339 focuses on a smaller set of date and time formats and has stricter rules. It doesn't include all the features of ISO 8601, like week dates or very long years. For a clear comparison, check this resource: [RFC 3339 vs. ISO 8601](https://ijmacd.github.io/rfc3339-iso8601/).

What do you think?

Reviewer: Is 3339 considered a simpler version of 8601?  I never thought that. I always thought they were just different. I was actually surprised when I discovered how much they overlap.

Author: RFC 3339 is a simplified and restricted subset of ISO 8601.

They have some similarities like Both formats support the basic date-time structure using the format `YYYY-MM-DDTHH:MM:SS` and some differences like ISO 8601 allows reduced precision in the time format (e.g, only specifying the hour: 14:30) and RFC 3339 requires the full `HH:MM:SS` format.

ISO 8601 provides more options than RFC 3339.

So, what should be our conclusion for this? Do you prefer one out of the above two options or you have anything else to add?

Reviewer: > RFC 3339 is a simplified and restricted subset of ISO 8601.

I have a problem with the word "subset".  It's not a subset.  There are 3339 formats that 8601 doesn't accept.  In the comparison, you can plainly see this.  It's not a subset because the 3339 circle did not exist completely within the 8601 circle.

Author: Then, I think a better way to express this is:

Dates and times are represented in [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6). While RFC 3339 shares many similarities with the date format also commonly known as [ISO8601 format](https://www.iso.org/iso-8601-date-and-time-format.html), they are distinct formats. RFC 3339 follows stricter rules and includes some formats that ISO 8601 does not accept. At the same time, RFC 3339 leaves out certain features of ISO 8601, such as week dates or very long years. For a clear comparison, refer to [RFC 3339 vs. ISO 8601](https://ijmacd.github.io/rfc3339-iso8601/).

Is it better now? Anything else to be included?

Reviewer: Now you have the opposite problem that "distinct" means there's no overlap.  Just say they're separate specs.  The RFC is by IETF, and the ISO is by, well, ISO.  They're different because they are attempts to standardize dates by two different standardization bodies.

Author: Dates and times are represented in RFC 3339, section 5.6. RFC 3339 and ISO 8601 are standards from different standardization bodies. RFC 3339 is from the IETF and ISO 8601 from the ISO. They are separate date format specifications with overlapping features. RFC 3339 enforces stricter rules and supports certain formats that ISO 8601 does not, while omitting features like week dates and very long years, which are part of ISO 8601. To explore their similarities and differences, refer to: RFC 3339 vs. ISO 8601.
--------------------------------------------------------
I have rephrased it as per your suggestion. I will add the link to IETF, ISO official site and RFC vs ISO.

Reviewer: I think we're getting lost in the weeds here.  We should **focus on RFC3339** and _maybe_ casually remark that, while it overlaps somewhat with ISO8601, it is a different specification with a different set of formats.

Reviewer 3: Why are we mentioning the ISO standard at all? It isn't actually relevant. There are innumerable other standards that it *isn't* defined by.

Reviewer: That's really what I'm trying to get at.

Reviewer: Still waiting on a resolution here.  I'm happy removing any mention of ISO8601.

- `"date-time"`: Date and time together, for example, `2018-11-13T20:20:39+00:00`.
- `"time"`: <StarInline label="New in draft 7" /> Time, for example, `20:20:39+00:00`.
- `"date"`: <StarInline label="New in draft 7" /> Date, for example, `2018-11-13`.

> - `"duration"`: <StarInline label="New in draft 2019-09" /> A duration as defined by the [ISO 8601 ABNF for "duration"](https://datatracker.ietf.org/doc/html/rfc3339#appendix-A). For example, `P3D` expresses a duration of 3 days.

Reviewer: We link to RFC3339 here, but say that it's from ISO8601.  I think the text is good, but maybe we should link to 8601 directly rather than to the 3339 appendix. 

The text in the spec is:

> The duration format is from the ISO 8601 ABNF as given in Appendix A of RFC 3339.

@mwadams thoughts?

Reviewer 2: Doesn't ISO keep their standards behind a paywall? Linking to ISO8601 may not be an option.

Reviewer: Oh, yeah.  That's dumb.  I guess that's why we copied everything relevant into that appendix.
